crates/fuel-gas-price-algorithm/src/v1.rs

```diff
@@ -515,60 +526,69 @@ impl AlgorithmUpdaterV1 {
-            Ok(())
-        }
+        let bytes_for_heights = self.drain_l2_block_bytes_for_heights(heights);
+        match bytes_for_heights {
+            Complete(bytes) => {
+                let new_cost_per_byte: u128 = recording_cost.checked_div(bytes).ok_or(
+                    Error::CouldNotCalculateCostPerByte {
+                        bytes,
+                        cost: recording_cost,
+                    },
+                )?;
+                let new_da_block_cost = self
+                    .latest_known_total_da_cost_excess
+                    .saturating_add(recording_cost);
+                self.latest_known_total_da_cost_excess = new_da_block_cost;
+                self.latest_da_cost_per_byte = new_cost_per_byte;
+            }
+            // If we weren't able to get the bytes for all the heights, we can't update the cost per byte in a logical way.
+            // We will just accept the predicted cost as the real cost and move on with the algorithm
+            // We can safely say that the cost didn't exceed the `recording_cost`, so we will take the
+            // minimum of the two costs
+            Incomplete(bytes) => {
+                let new_guess_recording_cost = bytes
+                    .saturating_mul(self.latest_da_cost_per_byte)
+                    .min(recording_cost);
+                let new_da_block_cost = self
+                    .latest_known_total_da_cost_excess
+                    .saturating_add(new_guess_recording_cost);
+                self.latest_known_total_da_cost_excess = new_da_block_cost;
+            }
+        };
+        Ok(())
     }
 
-    fn drain_l2_block_bytes_for_range(
-        &mut self,
-        height_range: Range<u32>,
-    ) -> Result<u128, Error> {
+    // Get the bytes for all specified heights, or get none of them.
+    // Always remove the blocks from the unrecorded blocks so they don't build up indefinitely
+    fn drain_l2_block_bytes_for_heights(&mut self, heights: &[u32]) -> BytesForHeights {
+        let mut missed_some_heights = false;
         let mut total: u128 = 0;
-        for expected_height in height_range {
-            let (actual_height, bytes) = self
-                .unrecorded_blocks
-                .pop_first()
-                .ok_or(Error::L2BlockExpectedNotFound(expected_height))?;
-            if actual_height != expected_height {
-                return Err(Error::L2BlockExpectedNotFound(expected_height));
+        for expected_height in heights {
+            let maybe_bytes = self.unrecorded_blocks.remove(expected_height);
+            if let Some(bytes) = maybe_bytes {
+                total = total.saturating_add(bytes as u128);
+            } else {
+                tracing::error!(
+                    "L2 block expected but not found in unrecorded blocks: {}",
+                    expected_height,
+                );
+                missed_some_heights = true;
             }
```

Comment:
If block is not part of the `self.unrecorded_blocks`, it means we already posted it to the DA in the past.

We could either just ignore it or only increase the known DA cost on the amount of fee that we've paid to return this amount.

Reply:
> or only increase the known DA cost on the amount of fee that we've paid to return this amount.

I don't know what this means. I'm not sure what `fee` you're talking about

Ignoring makes the most sense to me but maybe I'm missing something.

Reply:
If the block committer submits two bundles(or more) for the same block height, it means that we pay a fee to DA twice(or more). Maybe we want to include the cost of posting the second bundle into the known DA cost.

Reply:
Ah ah. Yeah. That is possible in theory. But probably only when finalization fails on L1, right? In that case we don't pay the first fee, but only the second.

This is another "improvement" I've discussed. Where we track "committed but not finalized" blocks, not just `unrecorded_blocks`. I've considered this something we'd like to do after we get the DA gas cost live.

Reply:
It is not additional "improvement", it is the case that you handle by this code=) Basically, it is only possible in the case described above. If `unrecorded_blocks` doesn't have some block, it is only possible if you already processed that block height in the past(so you had two or more bundles that covered this block height).

Based on your comments we can just do nothing for now, and cover this case later=)